### PR TITLE
Cluster image selection support

### PR DESCRIPTION
**Pull request by aleksejlopasov**
Now the user can select image version or provide custom image for the cluster.
Initialization-actions support has been added.
Secondary worker config has been added, though nodes are always preemptible.
Variable ALLOW_RANDOM_CLUSTER_NAMES has been added to Jupyterhub config.

### CODE REVIEW

**Review by JerryLeiDing — COMMENTED** [review]
A few minor text cleanups, otherwise this looks good

File: dataprocspawner/spawner.py [hunk 1]
```diff
@@ -1167,22 +1180,36 @@ def _apply_users_configs(self, cluster_data):
...
+            {
+              'executable_file': init_action
+            }
+          )
+
+    if self.user_options.get('image_version'):
+      if self.user_options.get('custom_image'):
+        config['master_config']['image_uri'] = self.user_options.get('custom_image')
+        config['worker_config']['image_uri'] = self.user_options.get('custom_image')
+        config['software_config']['image_version'] = \
```
Comment (JerryLeiDing) [inline dataprocspawner/spawner.py:1196]: Is this necessary? Dataproc's frontend should do the work of inferring the image version given a custom image URI, there's no reason for us to validate it here as well.

**Review by JerryLeiDing — COMMENTED** [review]
General note: it would be good to expand on the existing form unit test to ensure that all the options are being set as expected. The test can be found in tests/test_spawner.py on line 539.

File: dataprocspawner/spawner.py (hunk 1 above)
Comment (aleksejlopasov) [reply dataprocspawner/spawner.py:1196]: setting up the image version here is important if admins define different softwareConfig.imageVersion in YAML file.
Example: if 1.5-debian10 is set in YAML and custom image (1.5.17-debian10) is selected in UI, you will get an exception:
```
Error: HTTP 500: Internal Server Error (Error in Authenticator.pre_spawn_start:
InvalidArgument 400 Specified image version '1.5-debian10' doesn't match implied version '1.5.17-debian10' from uri 'https://www.googleapis.com/compute/v1/projects/<project-id>/global/images/<image-id>'!)
```
Additionally we check image_version for Component Gateway compatibility, so it has to be set up properly before the validation occurs.

File: dataprocspawner/customize_cluster.py [hunk 2]
```diff
@@ -152,7 +192,7 @@ def get_custom_cluster_html_form(autoscaling_policies, node_types):
 
   html_worker_disk_type = """
     <div class="form-group">
-      <label for="worker_disk_type">Worker disk type</label>
+      <label for="worker_disk_type">Primary disk type</label>
```
[the remark this answers is not in the document]
Comment (aleksejlopasov) [reply dataprocspawner/customize_cluster.py:195]: The new naming convention will be applied in next PR with Material UI

File: dataprocspawner/customize_cluster.py [hunk 3]
```diff
@@ -133,15 +169,19 @@ def get_custom_cluster_html_form(autoscaling_policies, node_types):
...
-  html_worker_type = ""
+  html_worker_type = """
+    <br />
+    <div class="form-group" name="worker">
+      <label for="worker"><h1>Worker nodes</h1></label>"""
   if node_types:
     html_worker_type += """
       <div class="form-group">
-        <label for="worker_node_type">Workers machine type</label>
+        <label for="worker_node_type">Machine type</label>
```
[the remark this answers is not in the document]
Comment (aleksejlopasov) [reply dataprocspawner/customize_cluster.py:184]: The new naming convention will be applied in next PR with Material UI

File: dataprocspawner/customize_cluster.py [hunk 4]
```diff
@@ -161,24 +201,60 @@ def get_custom_cluster_html_form(autoscaling_policies, node_types):
 
   html_worker_disk_size = """
     <div class="form-group">
-      <label for="worker_disk_size">Worker disk size</label>
+      <label for="worker_disk_size">Primary disk size</label>
```
[the remark this answers is not in the document]
Comment (aleksejlopasov) [reply dataprocspawner/customize_cluster.py:204]: The new naming convention will be applied in next PR with Material UI

File: dataprocspawner/customize_cluster.py [hunk 5]
```diff
@@ -161,24 +201,60 @@ def get_custom_cluster_html_form(autoscaling_policies, node_types):
...
+      <input name="worker_node_amount" class="form-control" placeholder="" value='2'></input>
+    </div>
+    </div>"""
+
+  html_secondary_worker_disk_type = """
+    <br />
+    <div class="form-group" name="sec_worker">
+      <label for="sec_worker"><h1>Secondary worker nodes</h1></label>
+      <div class="form-group">
+        <label for="sec_worker_disk_type">Primary disk type</label>
```
[the remark this answers is not in the document]
Comment (aleksejlopasov) [reply dataprocspawner/customize_cluster.py:220]: The new naming convention will be applied in next PR with Material UI

File: dataprocspawner/customize_cluster.py [hunk 6]
```diff
@@ -161,24 +201,60 @@ def get_custom_cluster_html_form(autoscaling_policies, node_types):
...
+        <select class="form-control" name="sec_worker_disk_type">
+          <option value="pd-standard">Standard Persistent Disk</option>
+          <option value="pd-ssd">SSD Persistent Disk</option>
+        </select>
+      </div>"""
+
+  html_secondary_worker_disk_size = """
+    <div class="form-group">
+      <label for="sec_worker_disk_size">Primary disk size</label>
```
[the remark this answers is not in the document]
Comment (aleksejlopasov) [reply dataprocspawner/customize_cluster.py:229]: The new naming convention will be applied in next PR with Material UI

File: dataprocspawner/spawner.py (hunk 1 above)
Comment (JerryLeiDing) [reply dataprocspawner/spawner.py:1196]: We can unset the image version if a custom image is set and vice versa.

The point about image version checking for Component Gateway is well taken. I'm a little concerned about latency implications for a call to GCE before spawning, but let's merge this and see if it actually has an impact.

**Review by JerryLeiDing — APPROVED** [review]

File: dataprocspawner/spawner.py [hunk 7]
```diff
@@ -1107,6 +1107,19 @@ def _validate_zones(self, region, zones):
...
+    image = custom_image.split('/')[4]
+    try:
+      request = self.compute_client.images().get(project=project, image=image)
+      response = request.execute()
+      version = re.split(r'(-debian.*|-ubuntu.*)', response['labels']['goog-dataproc-version'])
+      goog_version = version[0].replace('-', '.') + version[1]
+    except Exception as e: ## pylint: disable=broad-except
+      self.log.info(f'Failed to process Dataproc custom image version:\n\t{e}')
+
+    return goog_version
```
Comment (JerryLeiDing) [inline dataprocspawner/spawner.py:1121]: Oops, just realized goog_version needs to be initialized to empty string before the try block or else it'll blow up if an exception occurs.

Comment (aleksejlopasov) [reply dataprocspawner/spawner.py:1122]: sorry, my loss. fixed

**Pull request merged**